resource.go

```diff
@@ -79,16 +109,58 @@ var genericResourceMap = map[string]string{
 	"task_id":    stackdriverGenericTaskID,
 }
 
+var knativeRevisionResourceMap = map[string]string{
+	"project_id":             stackdriverProjectID,
+	"location":               resourcekeys.CloudKeyZone,
+	"cluster_name":           resourcekeys.K8SKeyClusterName,
+	knativeServiceName:       knativeServiceName,
+	knativeRevisionName:      knativeRevisionName,
+	knativeConfigurationName: knativeConfigurationName,
+	knativeNamespaceName:     knativeNamespaceName,
+}
+
+// getAutodetectedLabels returns all the labels from the Monitored Resource detected
+// from the environment by calling monitoredresource.Autodetect. If a "zone" label is detected,
+// a "location" label is added with the same value to account for differences between
+// Legacy Stackdriver and Stackdriver Kubernetes Engine Monitoring,
+// see https://cloud.google.com/monitoring/kubernetes-engine/migration.
+func getAutodetectedLabels() map[string]string {
```

Comment:
Can this be avoided by using MultiDetector as used in [auto.Detect()](https://github.com/census-ecosystem/opencensus-go-resource/blob/87ca538ec76b3961839940406845ef624d794fa8/auto/auto.go#L28-L33). Something like this in knative along with kantive-specific resource detector. Optionally, auto pkg could be enhanced to provide detector chain auto.Detectors().
```go
        resDetector := resource.MultiDetector(
		knative.Detect,
		gke.Detect,  // could be replaced by auto.Detectors()
		gcp.Detect,  //
		aws.Detect,  //
	)
	opts := []ocagent.ExporterOption{
		ocagent.WithResourceDetector(resDetector),
	}

	opts = append(opts, additionalOpts...)
	exp, err := ocagent.NewExporter(opts...)
```

The above would create generic labels for specific clouds and then SD. can map those to specific labels for its backend.

Reply:
For Knative, the need for any GCP metadata is solely to support exporting to Stackdriver, so I think it makes the most sense to put the fix in the Stackdriver exporter. 

Knative itself emphasizes portability and doesn't attempt to obtain cloud-specific metadata anywhere else. For example, with the current usage of the Prometheus exporter, which is the OSS default, Knative does not add any cloud-specific metric tags. Knative also doesn't support providing Google Application Default credentials or any similar interactions specific to Google Cloud APIs.

Also, autodetecting cloud labels on the Knative side also gets a bit messy if you're using a multi-cloud setup like sending metrics to Google's Stackdriver from AWS or VMware's Wavefront from GKE. Then the autodetected labels are extraneous and might have to be dropped at some point depending on how strict the metrics backend is, (though this is an edge case scenario).

Reply:
First, this will pull in dependencies on aws libs irrespective of user running it on aws or not.  See [this](https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/issues/35). Although you can argue that same would happen on knative if you take the approach I suggested.

Second, thing is it is probably not always desirable to auto-detect when a label is missing. What if the collector (where exporter is running) is part of a resource (missing one) that is different from resource of the source of the metric? In that case wouldn't it populate the  wrong interface? (It is corner case). Maybe there should be a configuration that enables auto-detect-on-missing-resource-labels.

Reply:
Hmm, I could do the same pkg re-factoring into monitoredresource/aws & monitoredresource/gcp then and only autodetect GCP labels?

I agree the edge case you mentioned could be confusing, but it seems better to send the metrics with the metrics collector's GCP labels than to drop the metric entirely. Plus, for embedded Stackdriver exporters, autodetecting would always be correct if successful.

I can still put it behind a configuration option though, though I would need to surface it on the OpenTelemetry collector service config too.

Reply:
Let's not have configuration now. If it is required in the future to disable the auto-detect then we can add it then.

Reply:
I split monitoredresource into monitoredresource/gcp and monitored/aws. I left the public surface of monitoredresource alone, but added comments that the structs are deprecated and the gcp or aws equivalents should be used instead.

I changed the defaultMapping Autodetection of labels to only be for GCP resources.